Firestore/core/test/unit/local/remote_document_cache_test.cc

```diff
@@ -230,34 +232,85 @@ TEST_P(RemoteDocumentCacheTest, DocumentsMatchingUsesReadTimeNotUpdateTime) {
         SetTestDocument("b/new", /* updateTime= */ 2, /* readTime= */ 1);
 
         core::Query query = Query("b");
-        DocumentMap results = cache_->GetMatching(query, Version(1));
-        std::vector<Document> docs = {
-            Doc("b/old", 1, kDocData),
+        MutableDocumentMap results = cache_->GetMatching(query, Version(1));
+        std::vector<MutableDocument> docs = {
+            Doc("b/old", 1, DeepClone(kDocData)),
         };
-        EXPECT_THAT(results.underlying_map(), HasExactlyDocs(docs));
+        EXPECT_THAT(results, HasExactlyDocs(docs));
       });
 }
 
+TEST_P(RemoteDocumentCacheTest, DoesNotApplyDocumentModificationsToCache) {
+  // This test verifies that the MemoryMutationCache returns copies of all
+  // data to ensure that the documents in the cache cannot be modified.
+  persistence_->Run("test_does_not_apply_document_modifications_to_cache", [&] {
+    MutableDocument document = SetTestDocument("coll/doc", Map("value", "old"));
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    MutableDocumentMap documents =
+        cache_->GetAll(DocumentKeySet{Key("coll/doc")});
+    document = documents.find(Key("coll/doc"))->second;
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    documents = cache_->GetMatching(Query("coll"), SnapshotVersion::None());
+    document = documents.find(Key("coll/doc"))->second;
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+  });
+}
 // MARK: - Helpers
 
-Document RemoteDocumentCacheTest::SetTestDocument(const absl::string_view path,
-                                                  int update_time,
-                                                  int read_time) {
-  Document doc = Doc(path, update_time, kDocData);
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path,
```

Comment:
Nit: I'd suggest omitting this `const` from `path` -- we generally avoid const parameters, so this one looks like it accidentally sneaked in.